crates/metaslang/bindings/src/resolver/c3.rs

```diff
@@ -0,0 +1,245 @@
+use std::collections::{HashMap, VecDeque};
+use std::fmt::{Debug, Display};
+use std::hash::Hash;
+
+/// Produces a linearisation of a hierarchy of items using the C3 linearisation
+/// algorithm. Given an item `A` with parents `(B1, B2)` in that order, the
+/// linearisation of A can be defined as:
+/// ```
+/// L(A) = [A] + merge(L(B1), L(B2), [B1, B2])
+/// ```
+/// The merge operation (defined below) will order the items of the vectors by
+/// taking candidates that can only appear in the first position in all of the
+/// vectors.
+/// NOTE: parents are considered left-to-right as in the traditional C3
+/// linearisation algorithm, ie. Python style.
+pub(crate) fn linearise<Item: Clone + Debug + Display + Eq + Hash + PartialEq>(
+    target: &Item,
+    parents: &HashMap<Item, Vec<Item>>,
+) -> Option<Vec<Item>> {
+    let mut linearisations: HashMap<Item, Vec<Item>> = HashMap::new();
+
+    // Keeps a running queue of pending linearisations.
+    let mut queue: VecDeque<Item> = VecDeque::new();
+    queue.push_back(target.clone());
+
+    // When an item cannot be resolved, we save it here along the number of
+    // items already linearised. If we fail to linearise a second time, we use
+    // this to check if progress was made to avoid infinite loops with cycles.
+    let mut checkpoint = None;
+
+    while let Some(item) = queue.pop_front() {
+        if linearisations.contains_key(&item) {
+            continue;
+        }
+        let item_parents = &parents[&item];
+        let mut merge_set = Vec::new();
+        for parent in item_parents {
+            if let Some(parent_linearisation) = linearisations.get(parent) {
```

Comment:
And similar to the above comment, I would use a `match`. Once again, just an observation, not a critique.

Reply:
Hmm... I was under the impression that using `match` here would trigger a clippy warning. But I must have confused it with other cases. I changed this and a bunch of other similar instances and I think it improves readability. Thanks!

Reply:
While changing other parts of the code, I found [a case](https://github.com/NomicFoundation/slang/blob/a8a847cf1fe757a3a8d5dd9211f13a9f349f3117/crates/solidity/outputs/cargo/tests/src/multi_part_file.rs#L56-L67) where clippy does emit a warning. Not exactly sure what is the differentiating factor.